prov/gni/src/gnix_ep.c

```diff
@@ -576,6 +588,8 @@ int gnix_ep_open(struct fid_domain *domain, struct fi_info *info,
 		}
 	}
 
-	if (nic != NULL) free(nic);
+	if (nic != NULL)
+		free(nic);
+	pthread_mutex_unlock(&gnix_nic_list_lock);
```

Comment:
I'm not clear on whether it would be wise to lock/unlock and have the list operations possibly interleaved.  In this particular case, to make the critical section smaller, you could move the unlocking to before each `goto`, but it's harder to maintain.